Defining a custom element should not cause global object's structure transitions

shvaikalesh opened the pull request:
#### 1b11bc2ebce8a49909b9396ce92bc947ec4d250c
<pre>
Defining a custom element should not cause global object&apos;s structure transitions
<a href="https://bugs.webkit.org/show_bug.cgi?id=233025">https://bugs.webkit.org/show_bug.cgi?id=233025</a>

Reviewed by Ryosuke Niwa.

This patch utilizes visitAdditionalChildren() to mark custom element&apos;s constructor / callbacks,
which saves up to 5 structure transitions per custom element definition.

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry-constructor-and-callbacks-are-held-strongly-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry-constructor-and-callbacks-are-held-strongly.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/resources/garbage-collect.js: Added.
(async maybeGarbageCollectAsync):
* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::visitJSFunctions):
* Source/WebCore/bindings/js/JSCustomElementInterface.h:
* Source/WebCore/bindings/js/JSCustomElementRegistryCustom.cpp:
(WebCore::JSCustomElementRegistry::define):
(WebCore::JSCustomElementRegistry::visitAdditionalChildren):
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::visitJSCustomElementInterfaces):
* Source/WebCore/dom/CustomElementRegistry.h:
* Source/WebCore/dom/CustomElementRegistry.idl:

Canonical link: <a href="https://commits.webkit.org/253330@main">https://commits.webkit.org/253330@main</a>
</pre>

rniwa commented on Source/WebCore/dom/CustomElementRegistry.h:
```diff
@@ -84,4 +85,11 @@ class CustomElementRegistry : public RefCounted<CustomElementRegistry>, public C
+template<typename Visitor>
+void CustomElementRegistry::visitJSCustomElementInterfaces(Visitor& visitor)
+{
+    for (auto* interface : m_constructorMap.values())
```
We need a lock since this function can be called concurrently while the main thread is running.

shvaikalesh replied:
Added, thank you!

rniwa reviewed (changes requested):
We need a lock around `m_constructorMap`.

rniwa commented on Source/WebCore/dom/CustomElementRegistry.h:
```diff
@@ -70,12 +71,14 @@ class CustomElementRegistry : public RefCounted<CustomElementRegistry>, public C
     DOMWindow& m_window;
     HashMap<AtomString, Ref<JSCustomElementInterface>> m_nameMap;
     HashMap<const JSC::JSObject*, JSCustomElementInterface*> m_constructorMap;
+    Lock m_visitConstructorMapLock;
```
We need to grab this lock whenever m_constructorMap is mutated.

rniwa commented on Source/WebCore/dom/CustomElementRegistry.h:
```diff
@@ -84,4 +87,12 @@ class CustomElementRegistry : public RefCounted<CustomElementRegistry>, public C
+template<typename Visitor>
+void CustomElementRegistry::visitJSCustomElementInterfaces(Visitor& visitor)
+{
+    Locker locker { m_visitConstructorMapLock };
```
This is useless if the mutator isn't doing the same.

rniwa reviewed (approved):
r=me.

Is there a WPT export somewhere?

shvaikalesh commented on Source/WebCore/dom/CustomElementRegistry.h:
```diff
@@ -81,6 +83,7 @@ class CustomElementRegistry : public RefCounted<CustomElementRegistry>, public C
     HashMap<const JSC::JSObject*, JSCustomElementInterface*> m_constructorMap;
     MemoryCompactRobinHoodHashMap<AtomString, Ref<DeferredPromise>> m_promiseMap;
     MemoryCompactRobinHoodHashSet<AtomString> m_disabledShadowSet;
+    Lock m_constructorMapLock;
```
I've temporarily removed `WTF_GUARDED_BY_LOCK` in attempt to figure out a bit weird Windows compilation error. Will make sure to bring back all the assertions once Windows builds. Fun.

shvaikalesh committed:
Defining a custom element should not cause global object's structure transitions
https://bugs.webkit.org/show_bug.cgi?id=233025

Reviewed by Ryosuke Niwa.

This patch utilizes visitAdditionalChildren() to mark custom element's constructor / callbacks,
which saves up to 5 structure transitions per custom element definition.

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry-constructor-and-callbacks-are-held-strongly-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/CustomElementRegistry-constructor-and-callbacks-are-held-strongly.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/resources/garbage-collect.js: Added.
(async maybeGarbageCollectAsync):
* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::visitJSFunctions):
* Source/WebCore/bindings/js/JSCustomElementInterface.h:
* Source/WebCore/bindings/js/JSCustomElementRegistryCustom.cpp:
(WebCore::JSCustomElementRegistry::define):
(WebCore::JSCustomElementRegistry::visitAdditionalChildren):
* Source/WebCore/dom/CustomElementRegistry.cpp:
(WebCore::CustomElementRegistry::visitJSCustomElementInterfaces):
* Source/WebCore/dom/CustomElementRegistry.h:
* Source/WebCore/dom/CustomElementRegistry.idl:

Canonical link: https://commits.webkit.org/253330@main
